features/FEATURE_LWIP/lwip-interface/lwip_stack.c

```diff
@@ -1123,6 +1154,106 @@ static nsapi_error_t mbed_lwip_setsockopt(nsapi_stack_t *stack, nsapi_socket_t h
+                return NSAPI_ERROR_PARAMETER;
+            }
+
+            /* Convert the interface address, or make sure it's the correct sort of "any" */
+            if (imr->imr_interface.version != NSAPI_UNSPEC) {
+                if (!convert_mbed_addr_to_lwip(&if_addr, &imr->imr_interface)) {
+                    return NSAPI_ERROR_PARAMETER;
+                }
+            } else {
+                ip_addr_set_any(IP_IS_V6(&if_addr), &if_addr);
+            }
+
+            igmp_err = ERR_USE; // Maps to NSAPI_ERROR_UNSUPPORTED
+            if(optname == NSAPI_ADD_MEMBERSHIP) {
+                #if LWIP_IPV4
+                if (IP_IS_V4(&if_addr)) {
+                    igmp_err = igmp_joingroup(ip_2_ip4(&if_addr), ip_2_ip4(&multi_addr));
+                }
+                #endif
+                #if LWIP_IPV6
+                if(IP_IS_V6(&if_addr)) {
+                    igmp_err = mld6_joingroup(ip_2_ip6(&if_addr), ip_2_ip6(&multi_addr));
+                }
+                #endif
+
+                if(igmp_err == ERR_OK) {
+                    if(!s->multicast_memberships) {
+                        // First multicast join on this socket, allocate space for membership tracking
+                        s->multicast_memberships = malloc(sizeof(nsapi_ip_mreq_t) * LWIP_SOCKET_MAX_MEMBERSHIPS);
+                    } else if(s->multicast_memberships_count == LWIP_SOCKET_MAX_MEMBERSHIPS) {
+                        return ERR_BUF;
+                    }
+
+                    uint32_t index = 0;
+                    next_free_multicast_member(s->multicast_memberships_registery, index);
+
+                    s->multicast_memberships[index].imr_multiaddr = imr->imr_multiaddr;
+                    s->multicast_memberships[index].imr_interface = imr->imr_interface;
+
+                    set_multicast_member_registry_bit(s->multicast_memberships_registery, index);
+                    s->multicast_memberships_count++;
+                }
+            } else {
+                #if LWIP_IPV4
+                if (IP_IS_V4(&if_addr)) {
+                    igmp_err = igmp_leavegroup(ip_2_ip4(&if_addr), ip_2_ip4(&multi_addr));
+                }
+                #endif
+                #if LWIP_IPV6
+                if(IP_IS_V6(&if_addr)) {
+                    igmp_err = mld6_leavegroup(ip_2_ip6(&if_addr), ip_2_ip6(&multi_addr));
+                }
+                #endif
+
+                if(igmp_err == ERR_OK) {
+                    uint32_t count = 0;
+                    uint32_t index = 0;
+                    // Set upper limit on while loop, should break out when the membership pair is found
+                    while(count < s->multicast_memberships_count) {
+                        next_registered_multicast_member(s->multicast_memberships_registery, index);
+
+                        bool address_matched = true;
+                        for(int i = 0; i < NSAPI_IP_BYTES; i++) {
```

Comment:
I'd be a bit less literal about this; rather than a manual loop, address_matched is just
```
memcmp(&s->multicast_memberships[index].imr_multiaddr, &imr->imr_multiaddr, sizeof(nsapi_addr_t)) == 0 &&
memcmp(&s->multicast_memberships[index].imr_interface, &imr->imr_interface, sizeof(nsapi_addr_t)) == 0
```
You could just memcmp the whole structure, if you assumed no padding, which I think is probably fine.